File: 234. Palindrome Linked List/Solution.py
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True

        fast = slow = head
        reverse_list = None
        while(fast is not None and fast.next is not None):
            pre = slow

            fast = fast.next.next
            slow = slow.next

            pre.next = reverse_list
            reverse_list = pre

            
        if fast is not None:
            slow = slow.next
            
        while reverse_list is not None and reverse_list.val == slow.val:
            reverse_list = reverse_list.next
            slow = slow.next

        return reverse_list is None
        

File: 234. Palindrome Linked List/Solution.java
public class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode fast = head;
        ListNode slow = head;
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }

        ListNode r = reverse(slow);

        while(r != null){
            if(r.val != head.val)return false;
            r = r.next;
            head = head.next;
        }

        return true;
    }

    public ListNode reverse(ListNode pointer){
        ListNode prev = null;
        while(pointer != null){
            ListNode next = pointer.next;
            pointer.next = prev;
            prev = pointer;
            pointer = next;
        }
        return prev;
    }
}


 class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }


